src/sdam/monitor.rs

```diff
@@ -184,7 +200,46 @@ impl HeartbeatMonitor {
         let result = match self.connection {
             Some(ref mut conn) => {
                 let command = is_master_command(self.client_options.server_api.as_ref());
-                run_is_master(command, conn).await
+                emit_event!(
+                    self,
+                    ServerHeartbeatStartedEvent {
+                        server_address: self.address.clone(),
+                    },
+                    handle_server_heartbeat_started_event
+                );
+
+                let start_time = Instant::now();
+                let response_result = conn.send_command(command, None).await;
+                let end_time = Instant::now();
+
+                let round_trip_time = end_time.duration_since(start_time);
+
+                match response_result {
+                    Ok(raw_response) => match raw_response.as_is_master_response(round_trip_time) {
+                        Ok(is_master_reply) => {
+                            emit_event!(
+                                self,
+                                ServerHeartbeatSucceededEvent {
+                                    duration: round_trip_time,
+                                    reply: raw_response.body::<Document>().unwrap_or_else(
+                                        |err| doc! { "deserialization error": err.to_string() }
+                                    ),
+                                    server_address: self.address.clone(),
+                                },
+                                handle_server_heartbeat_succeeded_event
+                            );
+                            Ok(is_master_reply)
+                        }
+                        Err(err) => {
+                            self.emit_heartbeat_failed_event(err.clone(), round_trip_time);
```

Comment:
We *can* borrow `self` here because we're finished with using the mutable borrow.